src/ashell/ide-webusb.c

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) 2017-2018, Intel Corporation.
+
+/**
+* @file
+* @brief WebUSB specific configuration file
+*
+* Provides WebUSB and Microsoft OS Descriptor 2.0 Platform Capability
+* descriptors. Also, it handles the WebUSB Device Requests and provides
+* the landing page and allowed origin descriptors.
+*/
+
+#include <stdio.h>
+#include <string.h>
+#include <device.h>
+#include <zephyr.h>
+
+#include "ide-comms.h"
+#include "ide-webusb.h"
+
+/* WebUSB Platform Capability Descriptor */
+static const u8_t webusb_bos_descriptor[] = {
+  /* Binary Object Store descriptor */
+  0x05, 0x0F, 0x1D, 0x00, 0x01,
+
+  /* WebUSB Platform Capability Descriptor:
+   * https://wicg.github.io/webusb/#webusb-platform-capability-descriptor
+   */
+  0x18, 0x10, 0x05, 0x00,
+
+  /* PlatformCapability UUID */
+  0x38, 0xB6, 0x08, 0x34, 0xA9, 0x09, 0xA0, 0x47,
+  0x8B, 0xFD, 0xA0, 0x76, 0x88, 0x15, 0xB6, 0x65,
+
+  /* Version, VendorCode, Landing Page */
+  0x00, 0x01, 0x01, 0x01,
+};
+
+/* URL Descriptor: https://wicg.github.io/webusb/#url-descriptor */
+static const u8_t webusb_origin_url[] = {
+  /* Length, DescriptorType, Scheme */
+  0x11, 0x03, 0x00,
+  'l', 'o', 'c', 'a', 'l', 'h', 'o', 's', 't', ':', '8', '0', '0', '0'
+};
+
+// Store the receive callback provided by the application here.
+static webusb_process_cb webusb_cb = NULL;
+// Method to reply when finished parsing, and ready for more data.
+static webusb_process_ack webusb_ack = NULL;
+
+/**
+ * @brief Custom handler for standard requests in
+ *        order to catch the request and return the
+ *        WebUSB Platform Capability Descriptor.
+ *
+ * @param pSetup    Information about the request to execute.
+ * @param len       Size of the buffer.
+ * @param data      Buffer containing the request result.
+ *
+ * @return  0 on success, negative errno code on fail
+ */
+int custom_handle_req(struct usb_setup_packet *pSetup,
+    s32_t *len, u8_t **data)
+{
+  if (GET_DESC_TYPE(pSetup->wValue) == DESCRIPTOR_TYPE_BOS) {
+    *data = (u8_t *)(&webusb_bos_descriptor);
+    *len = sizeof(webusb_bos_descriptor);
+
+    return 0;
+  }
+
+  return -ENOTSUP;
+}
+
+/**
+ * @brief Handler called for WebUSB vendor specific commands.
+ *
+ * @param pSetup    Information about the request to execute.
+ * @param len       Size of the buffer.
+ * @param data      Buffer containing the request result.
+ *
+ * @return  0 on success, negative errno code on fail.
+ */
+int vendor_handle_req(struct usb_setup_packet *pSetup,
+    s32_t *len, u8_t **data)
+{
+  /* Get URL request */
```

Comment:
this file should uses 4 spaces for indentation like others.

Reply:
Done.